### PR TITLE
Make vault run as the vault user instead of root

### PR DESCRIPTION
**Pull request by RRAlex**
Using the consul Dockerfile, I tried to use the same method to make vault run as a user.
setcap was installed and used to give vault the IPC_LOCK capability required to lock memory.

edit: https://www.vaultproject.io/docs/config/ mentions how to do the setcap I used :)

### CODE REVIEW

**Comment by jefferai** [comment]
Neat! I'll try this out at some point and we could get this in for the 0.6.2 container (and potentially backport it).

**Comment by RRAlex** [comment]
Tell me what you think of the patch and/or if you want any enhancements.
Cheers!

File: 0.6.1/Dockerfile [hunk 1]
```diff
@@ -8,8 +8,13 @@ ENV VAULT_VERSION=0.6.1
 # to provide HashiCorp-built versions of basic utilities like dumb-init and gosu.
 ENV DOCKER_BASE_VERSION=0.0.4
 
+# Create a vault user and group first so the IDs get set the same way,
+# even as the rest of this may change over time.
+RUN addgroup vault && \
+adduser -S -G vault vault
```
Comment (jefferai) [inline 0.6.1/Dockerfile]: Super pedantic: can you indent this to match the rest of the file?

File: 0.6.1/docker-entrypoint.sh [hunk 2]
```diff
@@ -61,4 +61,9 @@ elif vault --help "$1" 2>&1 | grep -q "vault $1"; then
     set -- vault "$@"
 fi
 
+# If we are running Vault, make sure it executes as the proper user.
+if [ "$1" = 'vault' ]; then
```
Comment (jefferai) [inline 0.6.1/docker-entrypoint.sh]: If you look in the Consul container (https://github.com/hashicorp/docker-consul/blob/master/0.X/docker-entrypoint.sh#L78) it makes sure permissions are correct if the config dir is bind-mounted in. I think that's probably a good idea -- what do you think?

Comment (RRAlex) [inline 0.6.1/docker-entrypoint.sh]: Indeed! Normalizing it like in Consul would be safer so not to break anything after user is switched, on current installations.

Comment (jefferai) [inline 0.6.1/docker-entrypoint.sh]: The problem is that your chown is in the Dockerfile, while in the Consul container I linked to it's in the entrypoint. So it happens at runtime *after* the bind mount to fix files that are bind mounted in.

Comment (RRAlex) [inline 0.6.1/docker-entrypoint.sh]: You're right, and a bind mount will reset the owner as root, but we're also mounting sub directories not /vault directly (from memory?). But to be safe, if there is no reasons for it, I should remove the chown.

As for chown to vault from the entrypoint, does vault has data there that needs modifications, or just read access?

Consul needs to write, but does vault?

In my case I only have a /vault/config/config.json, what are the other possibilities?

**Comment by jefferai** [comment]
@RRAlex I've pushed up 0.6.3 files -- can you make these changes against there instead? I don't want to change behavior of 0.6.1/0.6.2 after the fact. But ready to merge after that!

**Comment by jefferai** [comment]
Closing in favor of #13

**Pull request closed without merging**